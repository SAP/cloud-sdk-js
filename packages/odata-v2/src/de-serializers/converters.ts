/* eslint-disable valid-jsdoc */

import moment from 'moment';
import { Time, durationRegexV2 } from '@sap-cloud-sdk/odata-common/internal';

/**
 * @internal
 */
export function deserializeToTime(value: string): Time {
  const regexResult = durationRegexV2.exec(value);
  if (!regexResult) {
    throw new Error(`Failed to parse the value: ${value} to time.`);
  }
  const { hours, minutes, seconds } = regexResult?.groups || {};

  return {
    hours: hours ? parseInt(hours.replace('H', ''), 10) : 0,
    minutes: minutes ? parseInt(minutes.replace('M', ''), 10) : 0,
    seconds: seconds ? parseInt(seconds.replace('S', ''), 10) : 0
  };
}

/**
 * @internal
 */
export function serializeFromTime(value: Time): string {
  return (
    'PT' +
    leftpad(value.hours, 2) +
    'H' +
    leftpad(value.minutes, 2) +
    'M' +
    leftpad(value.seconds, 2) +
    'S'
  );
}

/**
 * @internal
 * This function can be used for both Edm.DateTime and and Edm.DateTimeOffset.
 */
export function deserializeToMoment(edmDateTime: string): moment.Moment {
  const dateTimeOffsetComponents =
    /^\/Date\((?<ticks>\d+)((?<sign>[+-])(?<offset>\d{4}))?\)\/$/.exec(
      edmDateTime
    )?.groups;
  if (!dateTimeOffsetComponents) {
    throw new Error(`Failed to parse edmDateTime: ${edmDateTime} to moment.`);
  }

  const timestamp = moment(parseInt(dateTimeOffsetComponents.ticks));

  if (dateTimeOffsetComponents.sign && dateTimeOffsetComponents.offset) {
    const offsetMultiplier = dateTimeOffsetComponents.sign === '+' ? 1 : -1;
    const offsetInMinutes = parseInt(dateTimeOffsetComponents.offset);
    return timestamp.utc().utcOffset(offsetMultiplier * offsetInMinutes);
  }

  return timestamp;
}

/**
 * @internal
 * This function can be used for both Edm.DateTime and and Edm.DateTimeOffset.
 */
export function serializeFromMoment(momentInstance: moment.Moment): string {
  const timestamp = momentInstance.unix() * 1000;

  // For some reason isUtc() returns wrong values here, so we use the internal flag directly
  if (momentInstance['_isUTC']) {
    const offset = Math.abs(momentInstance.utcOffset());
    const operator = momentInstance.utcOffset() >= 0 ? '+' : '-';
    return `/Date(${timestamp}${operator}${leftpad(offset, 4)})/`;
  }

  return `/Date(${timestamp})/`;
}

function leftpad(value: any, targetLength: number): string {
  const str = value.toString();
  if (str.length >= targetLength) {
    return str;
  }
  return '0'.repeat(targetLength - str.length) + str;
}
