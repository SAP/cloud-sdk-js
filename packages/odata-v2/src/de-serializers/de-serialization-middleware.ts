import BigNumber from 'bignumber.js';
import {
  DeSerializationMiddleware as DeSerializationMiddlewareBase,
  DeSerializer,
  EdmTypeCommon,
  ExclusiveEdmTypeV2,
  Time
} from '@sap-cloud-sdk/odata-common';
import moment from 'moment';

export interface DeSerializationMiddleware<
  BinaryT = string,
  BooleanT = boolean,
  ByteT = number,
  DecimalT = BigNumber,
  DoubleT = number,
  FloatT = number,
  Int16T = number,
  Int32T = number,
  Int64T = BigNumber,
  GuidT = string,
  SByteT = number,
  SingleT = number,
  StringT = string,
  AnyT = any,
  DateTimeT = moment.Moment,
  DateTimeOffsetT = moment.Moment,
  TimeT = Time
> extends DeSerializationMiddlewareBase<
    BinaryT,
    BooleanT,
    ByteT,
    DecimalT,
    DoubleT,
    FloatT,
    Int16T,
    Int32T,
    Int64T,
    GuidT,
    SByteT,
    SingleT,
    StringT,
    AnyT
  > {
  'Edm.DateTime': DeSerializer<DateTimeT>;
  'Edm.DateTimeOffset': DeSerializer<DateTimeOffsetT>;
  'Edm.Time': DeSerializer<TimeT>;
}

export type DeSerializationMiddlewareV2BASE = {
  [P in EdmTypeCommon | ExclusiveEdmTypeV2]: DeSerializer<any>;
};
