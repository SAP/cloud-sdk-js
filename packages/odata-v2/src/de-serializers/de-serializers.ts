import {
  DeSerializer,
  DeSerializers as DeSerializersCommon
} from '@sap-cloud-sdk/odata-common';

/**
 * Represents a set of functions that determine (de-)serialization per EDM type.
 */
export interface DeSerializers<
  BinaryT = any,
  BooleanT = any,
  ByteT = any,
  DecimalT = any,
  DoubleT = any,
  FloatT = any,
  Int16T = any,
  Int32T = any,
  Int64T = any,
  GuidT = any,
  SByteT = any,
  SingleT = any,
  StringT = any,
  AnyT = any,
  DateTimeT = any,
  DateTimeOffsetT = any,
  TimeT = any
> extends DeSerializersCommon<
    BinaryT,
    BooleanT,
    ByteT,
    DecimalT,
    DoubleT,
    FloatT,
    Int16T,
    Int32T,
    Int64T,
    GuidT,
    SByteT,
    SingleT,
    StringT,
    AnyT
  > {
  /**
   * DeSerializer for `Edm.DateTime` to the generic type `DateTimeT`.
   */
  'Edm.DateTime': DeSerializer<DateTimeT>;
  /**
   * DeSerializer for `Edm.DateTimeOffset` to the generic type `DateTimeOffsetT`.
   */
  'Edm.DateTimeOffset': DeSerializer<DateTimeOffsetT>;
  /**
   * DeSerializer for `Edm.Time` to the generic type `TimeT`.
   */
  'Edm.Time': DeSerializer<TimeT>;
}
