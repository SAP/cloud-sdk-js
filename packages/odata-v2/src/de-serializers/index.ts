export * from './converters';
export * from './custom-de-serializers';
export * from './de-serializers';
export * from './default-de-serializers';
export * from './entity-deserializer';
export * from './payload-value-converter';
