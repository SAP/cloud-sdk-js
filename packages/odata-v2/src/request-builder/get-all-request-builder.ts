import {
  EntityIdentifiable,
  GetAllRequestBuilderBase,
  ODataGetAllRequestConfig,
  Filterable,
  EntityApi,
  and
} from '@sap-cloud-sdk/odata-common/internal';
import { transformVariadicArgumentToArray } from '@sap-cloud-sdk/util';
import { Entity } from '../entity';
import {
  DefaultDeSerializers,
  DeSerializers,
  entityDeserializer
} from '../de-serializers';
import { createODataUri } from '../uri-conversion';
import { responseDataAccessor } from './response-data-accessor';

export class GetAllRequestBuilder<
    EntityT extends Entity,
    DeSerializersT extends DeSerializers = DefaultDeSerializers
  >
  extends GetAllRequestBuilderBase<EntityT, DeSerializersT>
  implements EntityIdentifiable<EntityT, DeSerializersT>
{
  /**
   * Creates an instance of GetAllRequestBuilder.
   * @param entityApi - Entity API for building and executing the request.
   */
  constructor(entityApi: EntityApi<EntityT, DeSerializersT>) {
    super(
      entityApi,
      new ODataGetAllRequestConfig(
        entityApi,
        createODataUri(entityApi.deSerializers)
      ),
      entityDeserializer(entityApi.deSerializers),
      responseDataAccessor
    );
  }

  /**
   * Add filter statements to the request.
   * @param expressions - Filter expressions to restrict the response.
   * @returns The request builder itself, to facilitate method chaining.
   */
  filter(expressions: Filterable<EntityT, DeSerializersT>[]): this;
  filter(...expressions: Filterable<EntityT, DeSerializersT>[]): this;
  filter(
    first:
      | undefined
      | Filterable<EntityT, DeSerializersT>
      | Filterable<EntityT, DeSerializersT>[],
    ...rest: Filterable<EntityT, DeSerializersT>[]
  ): this {
    this.requestConfig.filter = and(
      transformVariadicArgumentToArray(first, rest)
    );
    return this;
  }
}
