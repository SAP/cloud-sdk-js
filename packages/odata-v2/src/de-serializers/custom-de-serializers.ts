import {
  Time,
  CustomOrDefaultType as CustomOrDefaultTypeCommon
} from '@sap-cloud-sdk/odata-common';
import BigNumber from 'bignumber.js';
import { DeSerializers } from './de-serializers';
import {
  DefaultDeSerializers,
  defaultDeSerializers
} from './default-de-serializers';

/**
 * Get a complete set of (de-)serializers, that consists of the given partial custom (de-)serializers and default (de-)serializers (aka. default (de-)serializers merged with custom (de-)serializers).
 * The custom (de-)serializers are merged with the default (de-)serializers, while custom (de-)serializers take precedence.
 * @param customDeSerializers - Custom (de-)serialization functions.
 * @returns A full set of (de-)serialization functions, that combines custom and default (de-)serializers.
 */
export function mergeDefaultDeSerializersWith<
  BinaryT = string,
  BooleanT = boolean,
  ByteT = number,
  DecimalT = BigNumber,
  DoubleT = number,
  FloatT = number,
  Int16T = number,
  Int32T = number,
  Int64T = BigNumber,
  GuidT = string,
  SByteT = number,
  SingleT = number,
  StringT = string,
  AnyT = any,
  DateTimeT = moment.Moment,
  DateTimeOffsetT = moment.Moment,
  TimeT = Time
>(
  customDeSerializers: Partial<
    DeSerializers<
      BinaryT,
      BooleanT,
      ByteT,
      DecimalT,
      DoubleT,
      FloatT,
      Int16T,
      Int32T,
      Int64T,
      GuidT,
      SByteT,
      SingleT,
      StringT,
      AnyT,
      DateTimeT,
      DateTimeOffsetT,
      TimeT
    >
  >
): CustomDeSerializers<typeof customDeSerializers> {
  return {
    ...(defaultDeSerializers as any),
    ...(customDeSerializers || {})
  };
}

/**
 * Infers the deserialized type for an EDM type, based on custom (de-)serializers.
 * If the custom (de-)serializers specify a type for the given EDM type, this type is inferred.
 * Otherwise the given DefaultType is used.
 * @typeParam CustomDeSerializerT - Type of the custom (de-)serializers.
 * @typeParam EdmT - The EDM type to infer the type for.
 */
export type CustomOrDefaultType<CustomDeSerializerT, EdmT> =
  CustomOrDefaultTypeCommon<CustomDeSerializerT, EdmT, DefaultDeSerializers>;

/**
 * Type of the full set of (de-)serialization functions, that include custom (de-)serializers (aka. default (de-)serializers type merged with custom (de-)serializers type).
 */
export type CustomDeSerializers<T> = DeSerializers<
  CustomOrDefaultType<T, 'Edm.Binary'>,
  CustomOrDefaultType<T, 'Edm.Boolean'>,
  CustomOrDefaultType<T, 'Edm.Byte'>,
  CustomOrDefaultType<T, 'Edm.Decimal'>,
  CustomOrDefaultType<T, 'Edm.Double'>,
  CustomOrDefaultType<T, 'Edm.Float'>,
  CustomOrDefaultType<T, 'Edm.Int16'>,
  CustomOrDefaultType<T, 'Edm.Int32'>,
  CustomOrDefaultType<T, 'Edm.Int64'>,
  CustomOrDefaultType<T, 'Edm.Guid'>,
  CustomOrDefaultType<T, 'Edm.SByte'>,
  CustomOrDefaultType<T, 'Edm.Single'>,
  CustomOrDefaultType<T, 'Edm.String'>,
  CustomOrDefaultType<T, 'Edm.Any'>,
  CustomOrDefaultType<T, 'Edm.DateTime'>,
  CustomOrDefaultType<T, 'Edm.DateTimeOffset'>,
  CustomOrDefaultType<T, 'Edm.Time'>
>;
