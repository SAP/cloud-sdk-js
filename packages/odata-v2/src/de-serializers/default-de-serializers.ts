import {
  defaultDeSerializersRaw as defaultDeSerializersCommon,
  Time,
  wrapDefaultDeSerializers
} from '@sap-cloud-sdk/odata-common/internal';
import BigNumber from 'bignumber.js';
import {
  deserializeToMoment,
  deserializeToTime,
  serializeFromMoment,
  serializeFromTime
} from './converters';
import { DeSerializers } from './de-serializers';

/**
 * Type of the default (de-)serializers.
 */
export type DefaultDeSerializers = DeSerializers<
  string,
  boolean,
  number,
  BigNumber,
  number,
  number,
  number,
  number,
  BigNumber,
  string,
  number,
  number,
  string,
  any,
  moment.Moment,
  moment.Moment,
  Time
>;

/**
 * Default (de-)serializers without `null` and `undefined` handling.
 */
const defaultDeSerializersRaw: DefaultDeSerializers = {
  ...defaultDeSerializersCommon,
  'Edm.DateTime': {
    deserialize: deserializeToMoment,
    serialize: serializeFromMoment,
    serializeToUri: value =>
      `datetime'${value.toISOString().replace(/Z$/, '')}'`
  },
  'Edm.DateTimeOffset': {
    deserialize: deserializeToMoment,
    serialize: serializeFromMoment,
    serializeToUri: value => `datetimeoffset'${value.toISOString()}'`
  },
  'Edm.Time': {
    deserialize: deserializeToTime,
    serialize: serializeFromTime,
    serializeToUri: (value, serialize) => `time'${serialize(value)}'`
  },
  /* DeSerializers with v2 specific URI serializer defaults. */
  'Edm.Decimal': {
    ...defaultDeSerializersCommon['Edm.Decimal'],
    serializeToUri: (value, serialize) => `${serialize(value)}M`
  },
  'Edm.Guid': {
    ...defaultDeSerializersCommon['Edm.Guid'],
    serializeToUri: (value, serialize) => `guid'${serialize(value)}'`
  }
};

/**
 * The default (de-)serializers.
 */
export const defaultDeSerializers = wrapDefaultDeSerializers(
  defaultDeSerializersRaw
);
