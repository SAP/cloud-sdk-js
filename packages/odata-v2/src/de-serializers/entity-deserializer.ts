import {
  entityDeserializer as entityDeserializerBase,
  EntityDeserializer
} from '@sap-cloud-sdk/odata-common';
import { getLinkedCollectionResult } from '../request-builder/response-data-accessor';
import { extractODataEtag } from '../extract-odata-etag';
import { DeSerializers } from './de-serializers';

/**
 * Entity deserializer instance for v2 entities.
 * See {@link @sap-cloud-sdk/odata-common!EntityDeserializer} for the provided methods.
 * @param deSerializers - DeSerializer.
 * @returns EntityDeserializer.
 */
export function entityDeserializer<DeSerializersT extends DeSerializers>(
  deSerializers: DeSerializersT
): EntityDeserializer {
  return entityDeserializerBase(
    deSerializers,
    extractODataEtag,
    getLinkedCollectionResult
  );
}
