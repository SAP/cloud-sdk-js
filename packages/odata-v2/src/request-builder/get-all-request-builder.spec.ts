import nock from 'nock';
import * as httpClient from '@sap-cloud-sdk/http-client';
import { TestEntity } from '@sap-cloud-sdk/test-services-odata-v2/test-service';
import { encodeTypedClientRequest } from '@sap-cloud-sdk/http-client/dist/http-client';
import { asc, desc } from '@sap-cloud-sdk/odata-common';
import { wrapJwtInHeader } from '../../../connectivity/src/scp-cf/jwt';
import {
  defaultDestination,
  mockCountRequest,
  mockDestinationsEnv,
  mockGetRequest,
  unmockDestinationsEnv,
  createOriginalTestEntityData1,
  createOriginalTestEntityData2,
  expectAllMocksUsed,
  certificateMultipleResponse,
  certificateSingleResponse,
  mockInstanceDestinationsCall,
  mockServiceBindings,
  mockSingleDestinationCall,
  mockSubaccountDestinationsCall,
  onlyIssuerServiceToken,
  onlyIssuerXsuaaUrl,
  providerXsuaaUrl,
  providerServiceToken,
  createOriginalTestEntityDataWithLinks
} from '../../../../test-resources/test/test-util';
import { parseDestination } from '../../../connectivity/src/scp-cf/destination/destination';
import {
  testEntityApi,
  testEntitySingleLinkApi,
  createTestEntity,
  testEntityApiCustom,
  createTestEntityWithCustomDeSerializers
} from '../../test/test-util';
import { DefaultDeSerializers } from '../de-serializers';
import { GetAllRequestBuilder } from './get-all-request-builder';

describe('GetAllRequestBuilder', () => {
  let requestBuilder: GetAllRequestBuilder<TestEntity, DefaultDeSerializers>;

  afterEach(() => {
    unmockDestinationsEnv();
  });

  beforeEach(() => {
    requestBuilder = new GetAllRequestBuilder(testEntityApi);
  });

  describe('url', () => {
    it('should set ascending order', async () => {
      const expected =
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$orderby=ComplexTypeProperty/StringProperty%20asc';
      const request = await testEntityApi
        .requestBuilder()
        .getAll()
        .orderBy(asc(testEntityApi.schema.COMPLEX_TYPE_PROPERTY.stringProperty))
        .url(defaultDestination);
      expect(request).toBe(expected);
    });

    it('should set descending order', async () => {
      const expected =
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$orderby=ComplexTypeProperty/StringProperty%20desc';
      const request = await testEntityApi
        .requestBuilder()
        .getAll()
        .orderBy(
          desc(testEntityApi.schema.COMPLEX_TYPE_PROPERTY.stringProperty)
        )
        .url(defaultDestination);
      expect(request).toBe(expected);
    });

    it('should set ascending order as default if no order is specified', async () => {
      const expected =
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$orderby=ComplexTypeProperty/StringProperty%20asc';
      const request = await testEntityApi
        .requestBuilder()
        .getAll()
        .orderBy(testEntityApi.schema.COMPLEX_TYPE_PROPERTY.stringProperty)
        .url(defaultDestination);
      expect(request).toBe(expected);
    });

    it('should set the correct order when both default ascending order and descending order are passed', async () => {
      const expected =
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$orderby=ComplexTypeProperty/StringProperty%20asc,ComplexTypeProperty/DateTimeProperty%20desc';
      const stringProperty =
        testEntityApi.schema.COMPLEX_TYPE_PROPERTY.stringProperty;
      const dateTimeProperty =
        testEntityApi.schema.COMPLEX_TYPE_PROPERTY.dateTimeProperty;
      const request = await testEntityApi
        .requestBuilder()
        .getAll()
        .orderBy(stringProperty, desc(dateTimeProperty))
        .url(defaultDestination);
      expect(request).toBe(expected);
    });

    it('is built correctly', async () => {
      const expected =
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity';
      const actual = await requestBuilder.url(defaultDestination);
      expect(actual).toBe(expected);
    });

    it('is built correctly with URI encoding', async () => {
      const expected =
        "/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$filter=(StringProperty%20eq%20'%C3%A4%20%C3%B6%2B%20''c')";
      const actual = await requestBuilder
        .filter(testEntityApi.schema.STRING_PROPERTY.equals("ä ö+ 'c"))
        .url(defaultDestination);
      expect(actual).toBe(expected);
    });

    it('adds expand for nested selects', async () => {
      const expected =
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$select=to_SingleLink/BooleanProperty&$expand=to_SingleLink';
      const actual = await requestBuilder
        .select(
          testEntityApi.schema.TO_SINGLE_LINK.select(
            testEntitySingleLinkApi.schema.BOOLEAN_PROPERTY
          )
        )
        .url(defaultDestination);
      expect(actual).toBe(expected);
    });
  });

  describe('execute', () => {
    it('is possible to use untyped properties', async () => {
      const entityData1 = {
        SomethingTheSDKDoesNotSupport: 'SomeValue'
      };

      mockGetRequest(
        {
          query: { $select: 'SomethingTheSDKDoesNotSupport' },
          responseBody: { d: { results: [entityData1] } }
        },
        testEntityApi
      );

      const actual = await requestBuilder
        .select(testEntityApi.schema.SOMETHING_THE_SDK_DOES_NOT_SUPPORT)
        .execute(defaultDestination);
      expect(actual[0].somethingTheSdkDoesNotSupport).toBe('SomeValue');
    });

    it('returns all entities', async () => {
      const entityData1 = createOriginalTestEntityData1();
      const entityData2 = createOriginalTestEntityData2();

      mockGetRequest(
        {
          responseBody: { d: { results: [entityData1, entityData2] } }
        },
        testEntityApi
      );

      const actual = await requestBuilder.execute(defaultDestination);
      expect(actual).toEqual([
        createTestEntity(entityData1),
        createTestEntity(entityData2)
      ]);
    });

    it('top(1) returns the first entity', async () => {
      const entityData1 = createOriginalTestEntityData1();
      mockGetRequest(
        {
          query: { $top: 1 },
          responseBody: { d: { results: [entityData1] } }
        },
        testEntityApi
      );

      const actual = await requestBuilder.top(1).execute(defaultDestination);
      expect(actual).toEqual([createTestEntity(entityData1)]);
    });

    it('skip(1) skips the first entity', async () => {
      const entityData2 = createOriginalTestEntityData2();
      mockGetRequest(
        {
          query: { $skip: 1 },
          responseBody: { d: { results: [entityData2] } }
        },
        testEntityApi
      );
      const actual = await requestBuilder.skip(1).execute(defaultDestination);
      expect(actual).toEqual([createTestEntity(entityData2)]);
    });

    it('throws an error when the destination cannot be found', async () => {
      mockDestinationsEnv(defaultDestination);

      const getAllRequest = requestBuilder.execute({
        destinationName: 'NonExistentDestination'
      });

      await expect(getAllRequest).rejects.toThrowErrorMatchingSnapshot();
    });

    it('throws an error when request execution fails', async () => {
      mockGetRequest(
        {
          responseBody: { error: 'ERROR' },
          statusCode: 500
        },
        testEntityApi
      );

      const getAllRequest = requestBuilder.execute(defaultDestination);

      await expect(getAllRequest).rejects.toThrowErrorMatchingSnapshot();
    });

    it('considers custom timeout on the request', async () => {
      const entityData1 = createOriginalTestEntityData1();
      mockGetRequest(
        {
          query: { $top: 1 },
          responseBody: { d: { results: [entityData1] } },
          delay: 100
        },
        testEntityApi
      );

      try {
        await requestBuilder.top(1).timeout(10).execute(defaultDestination);
      } catch (err) {
        expect(err.cause.message).toBe('timeout of 10ms exceeded');
        return;
      }
      throw new Error('Should not reach here.');
    });

    it('sets custom headers instead of destination headers', async () => {
      const entityData = createOriginalTestEntityData1();
      const customAuthHeader = { Authorization: 'custom' };
      mockGetRequest(
        {
          headers: customAuthHeader,
          responseBody: { d: { results: [entityData] } }
        },
        testEntityApi
      );

      const destinationWithAuthHeader = {
        ...defaultDestination,
        Authorization: 'destination'
      };
      const actual = await requestBuilder
        .addCustomHeaders(customAuthHeader)
        .execute(destinationWithAuthHeader);

      expect(actual).toEqual([createTestEntity(entityData)]);
    });

    it('parses the raw number of count response', async () => {
      mockCountRequest(
        defaultDestination,
        4711,
        testEntityApi.requestBuilder().getAll()
      );
      const count = await requestBuilder.count().execute(defaultDestination);
      expect(count).toBe(4711);
    });

    it('executes request and deserializes entities with custom (de-)serializer', async () => {
      const entityData = createOriginalTestEntityDataWithLinks();

      mockGetRequest(
        {
          responseBody: { d: { results: [entityData] } }
        },
        testEntityApiCustom
      );

      const [entity] = await testEntityApiCustom
        .requestBuilder()
        .getAll()
        .execute(defaultDestination);
      expect(entity).toEqual(
        createTestEntityWithCustomDeSerializers(entityData)
      );
    });
  });

  describe('executeRaw', () => {
    it('returns request and raw response', async () => {
      const entityData1 = createOriginalTestEntityData1();
      const entityData2 = createOriginalTestEntityData2();
      const rawResponse = { d: { results: [entityData1, entityData2] } };

      mockGetRequest(
        {
          responseBody: rawResponse
        },
        testEntityApi
      );

      const actual = await requestBuilder.executeRaw(defaultDestination);
      expect(actual.data).toEqual(rawResponse);
      expect(actual.request.method).toBe('GET');
    });

    it('executes a request using the (iss) to build a token instead of a user JWT', async () => {
      mockServiceBindings();
      jest.clearAllMocks();

      const nocks = [
        nock(onlyIssuerXsuaaUrl)
          .post('/oauth/token')
          .times(1)
          .reply(200, { access_token: onlyIssuerServiceToken }),
        nock(providerXsuaaUrl)
          .post('/oauth/token')
          .times(1)
          .reply(200, { access_token: providerServiceToken }),
        mockInstanceDestinationsCall(nock, [], 200, onlyIssuerServiceToken),
        mockSubaccountDestinationsCall(
          nock,
          certificateMultipleResponse,
          200,
          onlyIssuerServiceToken
        ),
        mockSingleDestinationCall(
          nock,
          certificateSingleResponse,
          200,
          'ERNIE-UND-CERT',
          wrapJwtInHeader(onlyIssuerServiceToken).headers
        ),
        nock(certificateSingleResponse.destinationConfiguration.URL)
          .get(/.*/)
          .reply(200, 'iss token used on the way')
      ];
      const spy = jest.spyOn(httpClient, 'executeHttpRequest');
      const response = await requestBuilder.executeRaw({
        destinationName: 'ERNIE-UND-CERT',
        iss: onlyIssuerXsuaaUrl
      });
      expectAllMocksUsed(nocks);
      expect(spy).toHaveBeenCalledWith(
        parseDestination(certificateSingleResponse),
        {
          headers: {
            requestConfig: {
              accept: 'application/json',
              'content-type': 'application/json'
            }
          },
          parameterEncoder: encodeTypedClientRequest,
          params: {
            requestConfig: {}
          },
          url: 'sap/opu/odata/sap/API_TEST_SRV/A_TestEntity',
          method: 'get',
          data: undefined
        },
        { fetchCsrfToken: true }
      );
      expect(response.data).toBe('iss token used on the way');
    }, 60000);

    it('returns request and raw response count', async () => {
      mockCountRequest(
        defaultDestination,
        4711,
        testEntityApi.requestBuilder().getAll()
      );
      const actual = await requestBuilder
        .count()
        .executeRaw(defaultDestination);
      expect(actual.data).toEqual(4711);
      expect(actual.request.method).toBe('GET');
    });
  });
});
