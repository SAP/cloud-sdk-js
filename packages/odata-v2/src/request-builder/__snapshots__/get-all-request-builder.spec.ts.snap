// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAllRequestBuilder execute throws an error when request execution fails 1`] = `
"get request to /testination/sap/opu/odata/sap/API_TEST_SRV failed! 
\\"ERROR\\""
`;

exports[`GetAllRequestBuilder execute throws an error when the destination cannot be found 1`] = `"Could not find a destination with name \\"NonExistentDestination\\"! Unable to execute request."`;
